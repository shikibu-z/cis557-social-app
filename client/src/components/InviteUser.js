/* eslint-disable no-restricted-syntax */
/* eslint-disable react/jsx-no-bind */
/* eslint-disable react/jsx-props-no-spreading */
import React, { useEffect, useState } from 'react';
import { Modal, ListGroup, Button } from 'react-bootstrap';
import '../stylesheets/InviteUser.css';

const lib = require('../api/fetchers');

function InviteUser(props) {
  const id = Number(localStorage.getItem('currUser'));
  const { type, group } = props;
  const [members, setMembers] = useState([]);
  const [originalMembers, setOriginalMembers] = useState([]);
  const [memberQuery, setMemberQuery] = useState('');

  const [isLoading, setIsLoading] = useState(true);

  function handleAddMember(e) {
    const user = Number(e.target.getAttribute('data-index'));
    if (type === 'Add') {
      lib.addAdmin({ groupId: group, userId: user }).then(() => {
        const filteredArray = originalMembers.filter((obj) => obj.idUsers !== user);
        setOriginalMembers(filteredArray);
      });
    } else {
      lib.inviteUser({ senderId: id, receiverId: user, groupId: group });
      const filteredArray = originalMembers.filter((obj) => obj.idUsers !== user);
      setOriginalMembers(filteredArray);
    }
  }

  function renderMembers() {
    const mapRows = members.map((m) => (
      <ListGroup.Item key={m.idUsers} className="d-flex justify-content-between">
        <div>
          <img id="invite-user-photo" src={m.profilePhoto} alt="" />
          {m.username}
        </div>
        <Button data-index={m.idUsers} onClick={handleAddMember}>
          {type}
        </Button>
      </ListGroup.Item>
    ));
    return mapRows;
  }

  useEffect(() => {
    setMemberQuery('');
    setMembers(originalMembers);
  }, [originalMembers]);

  useEffect(async () => {
    let data = [];
    if (type === 'Add') {
      const res = await lib.getGroupMembers(group);
      const res1 = await lib.getAdmins(group);
      const admins = res1.data.map((a) => a.idUsers);
      for (const mem of res.data) {
        if (!admins.includes(mem.idUsers)) {
          data.push(mem);
        }
      }
    } else {
      const res = await lib.getFriends(id, group);
      if (res.status === 200) {
        data = res.data;
      }
    }
    setMembers(data);
    setOriginalMembers(data);
    setIsLoading(false);
  }, []);

  const handleFilterMembers = (event) => {
    const query = event.target.value;
    if (query !== '') {
      const result = originalMembers.filter(
        (m) => m.username.toLowerCase().startsWith(query.toLowerCase()),
      );
      setMembers(result);
    } else {
      setMembers(originalMembers);
    }
    setMemberQuery(query);
  };

  return (
    !isLoading
    && (
    <Modal {...props}>
      <Modal.Header closeButton />
      <Modal.Body>
        <input id="invite-friend-search" placeholder="Search a friend" value={memberQuery} onChange={handleFilterMembers} />
        <ListGroup>
          {renderMembers()}
        </ListGroup>
      </Modal.Body>
    </Modal>
    )
  );
}

export default InviteUser;
