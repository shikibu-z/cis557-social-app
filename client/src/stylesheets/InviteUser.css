#invite-user-photo {
    object-fit: cover;
    border-radius: 50%;
    width: 50px;
    height: 50px;
    margin: 5px;
}

#invite-friend-search {
    background: #FFFFFF;
    border-radius: 5px;
    width: 50%;
    height: 25px;
    margin-left: 25%;
    margin-bottom: 15px;
    text-align: center;
}