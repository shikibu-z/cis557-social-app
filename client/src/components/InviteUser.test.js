/*eslint-disable */
import React from 'react';
import renderer from 'react-test-renderer';
import InviteUser from './InviteUser';

test('Invite User Snapshot', () => {
  const component = renderer.create(<InviteUser />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
