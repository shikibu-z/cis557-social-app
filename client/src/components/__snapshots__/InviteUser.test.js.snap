// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invite User Snapshot 1`] = `null`;
